Changed demo alert to a status indicator and fixed missing JSDocs

Blackbaud-TrevorBurch opened the pull request.

blackbaud-ado commented:
Preview build at [https://host.nxt.blackbaud.com/skyux-forms?_sp.spa=pr-file-drop-docs-and-demo-updates](https://host.nxt.blackbaud.com/skyux-forms?_sp.spa=pr-file-drop-docs-and-demo-updates)

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-attachment.component.ts:
```diff
@@ -58,6 +58,9 @@ import {
 let uniqueId = 0;
 
+/**
+ * Provides an element to attach a single local file
```
```suggestion
 * Provides an element to attach a single local file.
```

Blackbaud-TrevorBurch replied:
Fixed.

blackbaud-johnly replied:
Looks like this suggestion wasn't actually committed. Mind if I go ahead and commit it?

Blackbaud-TrevorBurch replied:
I hadn't pushed the change. It is there now

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-attachment.component.ts:
```diff
@@ -89,24 +92,33 @@ export class SkyFileAttachmentComponent implements AfterViewInit, AfterContentIn
     return this._disabled;
```
The `disabled` input should specify a default of `false`.

And we also need to delete `This property accepts boolean values.` because `docs-tools` handles that now so that we don't have to call it out in the description.

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly replied:
I'm not actually seeing this change. When I follow the PR preview link, we don't have the default and still have the unnecessary sentence in the property description. Is something wonky going on here? It looks like all the checks have passed.

Blackbaud-TrevorBurch replied:
I hadn't pushed the change. I just did so you should have a new link soon.

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-attachment.component.ts:
```diff
@@ -89,24 +92,33 @@ export class SkyFileAttachmentComponent implements AfterViewInit, AfterContentIn
   public minFileSize: number = 0;
 
+  /**
+   * Specifies a custom validation function: `[validateFn]="validateFile"`.
```
What is this validating? Should we specify why users would want to provide a custom validation function?

Blackbaud-TrevorBurch replied:
I don't think we should try to specify why a user would use this. They basically would use this any time they need to check for something to do with the file that we don't check for (such as the file name being in a given format). The cases for this are probably too large to try to spell out and custom validation is something that is fairly standard with forms. This component is just kind of strange in how you have to do it. Normally this custom validation would be baked into the form but because of how this component works we have to use this input instead. 

I've updated the description here slightly. Let me know if you want to chat more about it.

blackbaud-johnly replied:
I defer to you on how much to say. I was just checking to see if we needed to provide more here. I made a couple suggestions to the wording, but feel free to disregard the suggestions if they don't work.

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-drop.component.ts:
```diff
@@ -25,6 +25,12 @@ import {
 } from './file-attachment.service';
 
 /**
+ * Provides an element to attach multiple files. It includes an option that users can select to
```
Can we tighten up the first two sentences to `Provides an element to attach multiple files where users can browse or drag and drop local files or provide hyperlinks to external files.`?

And in the last sentence about specifying custom content, how do consumers do that? We tell them that they can, but we don't tell them how. And none of the properties in the table that follow seem to be for custom content. Should we be more specific here about how they can do this? Or should we just delete this if it's not something we want to encourage?

Blackbaud-TrevorBurch replied:
I've used your suggestion for the beginning of the paragraph. As for the question about custom content - I honestly kind of like what we have. It is hard to be more specific than that...... as to if we want to say it at all - I feel like we probably should say it as it is possible. We don't show doing this in the demo though. Maybe we file a separate issue to add a code example for this? I don't know if we'd want to do that or not - we'd probably need Todd's input on that.

blackbaud-johnly replied:
Not sure if it's worth following an issue. I was just making sure our docs are sufficient. I'll leave it up to you on whether this is worth following up on.

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-drop.component.ts:
```diff
@@ -25,6 +25,12 @@ import {
+ * to provide hyperlinks to attach external files. You can leave the contents of the component
+ * blank to display the drop zone's default UI, or you can specify custom content to
+ * display instead.
+ *
```
```suggestion
```
In the preview link, the second paragraph does not display. I'm not sure if it's this blank line in between, but I suspect that it might have something to do with it. I'm not sure that JSDocs allows for paragraph breaks. I'd just delete the blank line and make this all one paragraph.

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-drop.component.ts:
```diff
@@ -25,6 +25,12 @@ import {
+ * blank to display the drop zone's default UI, or you can specify custom content to
+ * display instead.
+ *
  * When the SKY UX module initializes, it disables the ability to drag and drop files
```
```suggestion
 * When the module initializes, it disables the ability to drag and drop files
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-drop.component.ts:
```diff
@@ -25,6 +25,12 @@ import {
+ * display instead.
+ *
  * When the SKY UX module initializes, it disables the ability to drag and drop files
  * for the entire window. This prevents the browser from opening files that are accidentally
```
```suggestion
 * for the entire window to prevent the browser from opening files that are accidentally
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-drop.component.ts:
```diff
@@ -53,13 +62,13 @@ export class SkyFileDropComponent implements OnDestroy {
 
   /**
-   * 	Specifies an accessibility label to provide a text equivalent for screen readers for the file upload button.
+   * Specifies an accessibility label to provide a text equivalent for screen readers for the file upload button.
```
```suggestion
   * Specifies an accessibility label for the file upload button to provide a text equivalent for screen readers.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-drop.component.ts:
```diff
@@ -53,13 +62,13 @@ export class SkyFileDropComponent implements OnDestroy {
 
   /**
-   * 	Specifies an accessibility label to provide a text equivalent for screen readers for the link upload input.
+   * Specifies an accessibility label to provide a text equivalent for screen readers for the link upload input.
```
```suggestion
   * Specifies an accessibility label for the link upload input to provide a text equivalent for screen readers.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/types/file-attachment-change.ts:
```diff
@@ -3,5 +3,8 @@ import {
 
 export interface SkyFileAttachmentChange {
+  /**
+   * Specifies the file which has been added or removed.
```
```suggestion
   * Specifies the file that was added or removed.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/types/file-attachment-click.ts:
```diff
@@ -3,5 +3,8 @@ import {
 
 export interface SkyFileAttachmentClick {
+  /**
+   * Specifies the file which has been clicked.
```
```suggestion
   * Specifies the file that was clicked.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/types/file-drop-change.ts:
```diff
@@ -3,6 +3,12 @@ import {
 
 export interface SkyFileDropChange {
+  /**
+   * Specifies an array of files which have been added or removed.
```
```suggestion
   * Specifies an array of files that were added or removed.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/types/file-drop-change.ts:
```diff
@@ -3,6 +3,12 @@ import {
+   */
   files: Array<SkyFileItem>;
+  /**
+   * Specifies an array of files which have been rejected.
```
```suggestion
   * Specifies an array of files that were rejected.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-item.ts:
```diff
@@ -3,8 +3,20 @@ import {
 
 export interface SkyFileItem {
+  /**
+   * The object which has been added or removed.
```
```suggestion
   * The object that was added or removed.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-item.ts:
```diff
@@ -3,8 +3,20 @@ import {
+   */
   file: File;
+  /**
+   * The data URL for the file which has been added or removed.
```
```suggestion
   * The data URL for the file that was added or removed.
```

Also, why do we have `data` before `URL` here but not in the description under `SkyFileLink`? I assume we should delete it here unless it is this adding important information. If so, we should we add `data` to other instances of `URL` in these docs, shouldn't we?

Blackbaud-TrevorBurch replied:
Nope. This actually was an intended difference. Data URLs, in a nutshell, are encoded versions of the file.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs

blackbaud-johnly replied:
Is that common knowledge for developers, or should we make "data URL" a link to the URL you shared?

Also, can we make the suggested tweak to the wording?

Blackbaud-TrevorBurch replied:
Oops. Totally missed that suggestion. Done and also added that link - I don't think it hurts anything to be specific there.

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-item.ts:
```diff
@@ -3,8 +3,20 @@ import {
+   */
   url: string;
+  /**
+   * The type of error which caused the file to be rejected.
```
```suggestion
   * The type of error that caused the file to be rejected.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-item.ts:
```diff
@@ -3,8 +3,20 @@ import {
+   */
   errorType: SkyFileItemErrorType;
+  /**
+   * Additional paramaters regarding the error which caused the file to be rejected.
```
```suggestion
   * Additional parameters about the error that caused the file to be rejected.
```

Blackbaud-TrevorBurch replied:
Done

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-link.ts:
```diff
@@ -1,3 +1,6 @@
 export interface SkyFileLink {
+  /**
+   * Specifies the URL for the file being linked.
```
```suggestion
   * Specifies the URL for the linked file.
```

Blackbaud-TrevorBurch replied:
Done

Blackbaud-TrevorBurch commented:
@blackbaud-johnly - I think I've addressed everything at this point. I did make a small tweak to `acceptedTypes` on both components as we were basically showing how to use an input. I didn't think that was necessary so I removed that portion of the description.

blackbaud-ado commented:
Preview build at [https://host.nxt.blackbaud.com/skyux-forms?_sp.spa=pr-file-drop-docs-and-demo-updates](https://host.nxt.blackbaud.com/skyux-forms?_sp.spa=pr-file-drop-docs-and-demo-updates)

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-attachment.component.ts:
```diff
@@ -89,24 +93,34 @@ export class SkyFileAttachmentComponent implements AfterViewInit, AfterContentIn
   public minFileSize: number = 0;
 
+  /**
+   * Specifies a custom validation function. This validation is ran along with all internal
```
```suggestion
   * Specifies a custom validation function. This validation runs alongside the internal
```

Blackbaud-TrevorBurch replied:
Done - both in this component and the file drop component.

blackbaud-johnly commented on src/app/public/modules/file-attachment/file-attachment.component.ts:
```diff
@@ -89,24 +93,34 @@ export class SkyFileAttachmentComponent implements AfterViewInit, AfterContentIn
 
+  /**
+   * Specifies a custom validation function. This validation is ran along with all internal
+   * file validation. This function should take a `SkyFileItem` in as a parameter.
```
```suggestion
   * file validation. This function takes a `SkyFileItem` interface as a parameter.
```

Blackbaud-TrevorBurch replied:
Hmmm - I know that technically that `SkyFileItem` is an interface; however, the function will take in an instance of that interface so I don't love saying that it takes in the interface. How about this?

"This function takes a `SkyFileItem` object as a parameter."

I've made my suggested change to both this and the file drop component but we can tweak it again if needed.

blackbaud-johnly replied:
Sounds good. Sorry I got it wrong. I was just trying to avoid `should` to avoid sounding uncertain, and it felt like we needed something after `SkyFileItem`. If `object works better than `interface`, that works for me. Thanks!

blackbaud-ado commented:
Preview build at [https://host.nxt.blackbaud.com/skyux-forms?_sp.spa=pr-file-drop-docs-and-demo-updates](https://host.nxt.blackbaud.com/skyux-forms?_sp.spa=pr-file-drop-docs-and-demo-updates)
